README.md
Toner
=====

"Toner" is the name of the B+W (black and white) map tiles, and Mapnik stylesheets, used for the
Dotspotting project. Toner was designed by [Geraldine Sarmiento](http://stamen.com/studio/geraldine).

Examples
--

https://github.com/Citytracking/toner/tree/master/examples

How to use Toner
--

At the end of the day Toner generates a stylesheet that can be used by a program
called Mapnik to draw maps. Those maps might be a single large image for print
or a lot of small images for map tiles but as far as Mapnik is concerned there
is an image of a set size that covers a specific geographic area and a bunch of
rules (styles) for how to draw the stuff inside those boxes.

Toner relies on a tool called Cascadenik so that map styles can be defined using
a CSS-like syntax that is a little more friendly than the XML-based markup
language that Mapnik uses by default. With Cascadenik you define two kinds of
files: Things ending in '.mss' are where the actual look and feel for a map;
Things ending in '.mml' are where you define administrative bits like database
passwords and queries for things to show on the map.

This is the place where are the non-database-y things required to generate the
"Toner" Mapnik stylesheets are kept.

The style.mml and related .mss files control the look and feel of the Toner 
cartography. These files are writen in a preprocessor language called Cascadenik. 
If you make changes to the stylesheets, you will need to convert this to the native
XML format that Mapnik supports to render the map tiles. 

Use the following command (from inside the 'mapnik' directory):

	make style.xml

Which is really just a shortcut for typing this:

	cascadenik-compile.py style.mml > style.xml

This will take your 'style.mml' and all the '.mss' files (the things that define
the look and feel of the Toner maps) and smush them together to create a new
file called 'style.xml'. This new file is what Mapnik uses to render maps.

For example, once you've created your new 'style.xml' file you could use the
'nik2img' program that is included with Mapnik to render a map of the USA like
this:

	nik2img.py -b -126 24 -66 49 -s 900913 -d 1024 768 \
		--fonts=fonts/Arial.ttf,fonts/Arial\ Bold.ttf,fonts/Arial\ Italic.ttf \
		style.xml usa.png

A sample 'usa.png' rendering is included in the 'examples' directory.

If you wanted to use the TileStache server libraries to draw map tiles using
Mapnik, you would create a new layer (in your TileStache config file) like this:

	"toner": {
		"provider": {
			"name": "mapnik",
			"mapfile": "/path/to/mapnik/style.xml",
			"fonts": "/path/to/mapnik/fonts"
		},
		"projection": "spherical mercator",
		"metatile": { "rows": 4, "columns": 4 }
	}

Take a look in the 'tilestache' directory for more details about using the
Toner stylesheets with TileStache.


Make map sandwitches
--

* **More Toner flavors**: Introduces specific flavors of Toner optimized for [map sandwiches](http://blogs.esri.com/esri/arcgis/2009/07/13/the-map-sandwich/), easy to integrating with and promoting your custom map stories: toner-standard ([toner](tilefarm.stamen.com/toner-2012)), toner-hybrid-with-labels ([toner-hybrid](tilefarm.stamen.com/toner-hybrid-with-labels)), toner-hybrid-only-lines ([toner-lines](tilefarm.stamen.com/toner-lines)), toner-hybrid-only-labels ([toner-labels](tilefarm.stamen.com/toner-labels)), toner-no-labels ([toner-background](tilefarm.stamen.com/toner-no-labels)).

Dependencies
--

The short version is: There are a lot.

The long version is: The gritty details of installing some of the tools that
Toner uses are outside the scope of this document. We've tried to give you the
shape of what you need to do and linked to the available documentation elsewhere.

Software Dependencies (required)
--

* A PostGIS database (http://postgis.refractions.net/)

* The osm2pgsql application for importing OpenStreetMap in to PostGIS (http://wiki.openstreetmap.org/wiki/Osm2pgsql)

* Python 2.6 or higher (http://www.python.org/)

* The GDAL libraries and utlities (http://www.gdal.org/)

* Mapnik and the Mapnik Python bindings (http://mapnik.org/)

* The ModestMaps Python libraries (http://pypi.python.org/pypi/ModestMaps/)

* The Cascadenik Python libraries (http://pypi.python.org/pypi/cascadenik/)

Software Dependencies (optional)
--

* The TileStache Python libraries (http://pypi.python.org/pypi/TileStache/)

* The gunicorn WSGI web server framework (http://www.gunicorn.org/)

* The pyproj Python libraries, required for using using the 'mapnik-render.py' script (http://pypi.python.org/pypi/pyproj)

* The Python Imaging Libraries, required for using the 'mapnik-render.py' script (http://pypi.python.org/pypi/PIL/)

Data Dependencies (required)
--

* PostGIS database tables for OpenStreetMap in EPSG:900913 (http://wiki.openstreetmap.org/wiki/PostGIS/)

* PostGIS database tables for the OpenStreetMap coastline in EPSG 90091 (http://tile.openstreetmap.org/processed_p.tar.bz2)

* PostGIS database tables for Natural Earth in EPSG:900913 (http://www.naturalearthdata.com/)

* NEW! PostGIS database tables for beta 1.5 Natural Earth roads in EPSG:900913 (included here). Use the included import script.

* NEW! PostGIS database tables for City labels and townspots in EPSG:900913 (included here). Use the included import script.


Database Tables
--

The Toner stylesheets need access to a pair of PostGIS databases:

* A database containing PlanetOSM data (as created by the osm2pgsql script) and
  coastline using the spherical mercator projection (EPSG:900913). See below for
  details.

* A database containing NaturalEarth data, as created by the shp2pgsql script
  (this is installed with PostGIS), using the spherical mercator projection
  (EPSG:900913). See below for details.

OpenStreetMap (OSM)
--

OSM publishes freely available downloads of their entire dataset at
http://planet.openstreetmap.org. Instructions for installing and setting up OSM
are outside the scope of this document but the OSM site has thorough
documentation available at: http://wiki.openstreetmap.org/wiki/PostGIS

If you don't want to install the entire OSM planet database but want to render
tiles for a smaller area you can also use the MirrorOSM tile provider in
TileStache to retrieve and store OSM data in PostGIS. Details are available over
here:

http://www.tilestache.org/doc/TileStache.Goodies.Providers.MirrorOSM.html

You will also need to add a copy of the OSM coastline to your planet_osm
database. The OSM coastline is distributed as a shapefile that you will need to
import using the 'shp2pgsql' program:

http://tile.openstreetmap.org/processed_p.tar.bz2

OSM-related
--

Toner uses a table containing OSM-derived data called
'planet_osm_motorways'. There's a handy PGSQL script called 'motorways.pgsql' in
the 'osm' table that you can run (once you've set up your planet_osm tables) to
create the new table.

Natural Earth
--

NaturalEarth is a public domain map dataset of various cultural and vector
datasets. It is available for download at: http://www.naturalearthdata.com/

Toner uses many, but not all of the datasets in NaturalEarth so the easiest
thing is just to grab the shapefiles we use as a single compressed file over
here:

http://citytracking.s3.amazonaws.com/toner/toner-naturalearth-1.1-epsg900913.zip

This file contains the 13 NaturalEarth datasets Toner uses as shapefiles
projected using EPGS:900913 (sometimes known as "spherical mercator" which
really just means "good for making map tiles"). They also correct a known issue
with the NaturalEarth shapefiles where polygons crossing the 180° meridian start
to behave badly.

You will still need to use the 'shp2pgsql' program to import them in to your
PostGIS database.

If you want to install of all NaturalEarth from scratch there are a few things
you should be aware of first.

The various datasets are available as separate downloads but here's a simple
shell script that will download them all, one at a time:

https://github.com/straup/naturalearth-tools/blob/master/fetchall-vector.sh

NaturalEarth is distributed using the WGS84 map projection so you will need to
reproject all the various shapefiles (in to the spherical mercator projection)
before importing them in to your database.

Here's another simple shell script that wraps up the entire process of
converting the NaturalEarth shapefiles and then importing them in to a PostGIS
database. It assumes that you have installed all the dependencies listed above:

https://github.com/straup/naturalearth-tools/blob/master/shp2pgmerc.sh

Included data
--

https://github.com/citytracking/toner/blob/master/import_included_map_data_to_postgis.sh

We've prepped city labels with [Dymo](https://github.com/migurski/dymo) at several
zoom levels so you don't have to (it takes weeks). The resulting shapefiles (SHP and
related extensions) should be imported into PostGIS for optimal performance and we 
include a script for that. The MML file assumes this step has been accomplished.

Also included are a dump of beta 1.5 Natural Earth roads for the mid-zooms. The import
script pushes those into PostGIS, too.

Other Stuff
--

https://github.com/straup/postgis-tools

These are mostly just a bunch of vanilla shell scripts that handle the multi-step
process of setting up a PostGIS database.


CHANGELOG.md
**Changelog**

_Published by YYYY. If second or more updates in a year, YYYY-MM. Milestones noted with each publication date._


# Toner 2012

_[Preview Toner 2012 »](http://tile.stamen.com/toner-2012/)._

***About Toner v2 aka 2012, "Milestone 2" in Github***

![Toner changes animated thru time in San Francisco](https://github.com/Citytracking/toner/raw/master/images/toner_animated_gif_v1_2010_to_v2_gm_2012.gif)

## General

* **Easy-to-use tiles**: Stamen now hosts easy to embed Toner tiles with CC license from [maps.stamen.com](http://maps.stamen.com), thanks to the [Knight Foundation](http://newschallenge.tumblr.com/) and our [Citytracking.org](http://citytracking.org/) [grant](http://content.stamen.com/knight_news_challenge_update)! No server hardware or software setup needed, just start using the tiles in your favorite web mapping API client side. You can still roll your own tiles using the data and setup readme's in the Github repo. [Read more »](http://maps.stamen.com)

* **More international**: Plays better outside of the United States! Now displays local names in non-Latin writing scripts (like Japanese and Arabic) and better accent marks in Europe. We optimized the road symbology to more places world wide. [Issue 30](https://github.com/Citytracking/toner/issues/30)

* **More Toner flavors**: Introduces specific flavors of Toner optimized for [map sandwiches](http://blogs.esri.com/esri/arcgis/2009/07/13/the-map-sandwich/), easy to integrating with and promoting your custom map stories: toner-standard ([toner](tilefarm.stamen.com/toner-2012)), toner-hybrid-with-labels ([toner-hybrid](tilefarm.stamen.com/toner-hybrid-with-labels)), toner-hybrid-only-lines ([toner-lines](tilefarm.stamen.com/toner-lines)), toner-hybrid-only-labels ([toner-labels](tilefarm.stamen.com/toner-labels)), toner-no-labels ([toner-background](tilefarm.stamen.com/toner-no-labels)). [Issue 10](https://github.com/Citytracking/toner/issues/10).

* **Easy to read stylesheets**: General stylesheet cleanup, consolidation. Restructured all the OSM roads using [High Roads](http://github.com/migurski/high-roads/). Now uses Postgres views by zoom level, making it much easier to design what big, medium, and small roads should look like consistently between layers while abstracting the data part. Similar appraoch is taken for water bodies using [Imposm](http://imposm.org/) tables. [Issue 9](https://github.com/Citytracking/toner/issues/9)

* **More content**: Added reservoirs, state boundaries, and more. Map now zooms to 19+, important when you're inventory mapping stories at the city block level where locations along a street and buildings/venues are helpful. Before they stopped at zoom 18 but often when you're looking at street-level incidents (as in [Dotspotting.org](http://www.dotspotting.org)), you need more detail [Issue 18](https://github.com/Citytracking/toner/issues/18). Along with that, the transition between bold black roads and cased white roads now starts at zoom 18 and carries thru to zoom 19+ ([Issue 17](https://github.com/Citytracking/toner/issues/17)). This preserves the strong contrast of Toner, but also allows better use as a background map visually at these zooms so your story points stay the focus (and consumes much less ink if you print the maps using a service like Stamen's [Field Papers](http://www.fieldpapers.org). 

* **Urban wayfinding**: At these most detailed zooms building footprints help us orient to the build landscape so we start adding those progressively in starting around zoom 14 (big airport terminals and convention center sized buildings) but most noticeably at zoom 16+. We also show metro (subway) stations now, helping navigate by landmarks in big cities like New York, London, and Tokyo. [Issue 16](https://github.com/Citytracking/toner/issues/16),  [Issue 48](https://github.com/Citytracking/toner/issues/48), [Issue 40](https://github.com/Citytracking/toner/issues/40).

## Boundaries

* **Improved graphic styling of country boundary lines** at zooms 8 and 9. [Issue 27](https://github.com/Citytracking/toner/issues/27)

* **Added state boundary lines** at the city and regional zooms. Important for places like Washington DC where a metropolitan area sprawls across multiple admin-1 jurisdictions. Made sure they stack above the water and made upstream changes in OSM master data to allow for boudnaries in the water that aren't indicator level to be not shown in Toner when using newest OSM [Issue 11](https://github.com/Citytracking/toner/issues/11), [Issue 24](https://github.com/Citytracking/toner/issues/24), [Issue 7](https://github.com/Citytracking/toner/issues/7), [Issue 6](https://github.com/Citytracking/toner/issues/6), and [Issue 50](https://github.com/Citytracking/toner/issues/50).

## Map labels

* Added support for **international Unicode (UTF-8) labels** from OSM by re-authoring fonts. Primarily seen in street labels and park names. [Issue 30](https://github.com/Citytracking/toner/issues/30)

* **Removed map label overlap** by manually adjusting the Dymo output around other map features like bodies of waters, country labels, and state labels. [Issue 34](https://github.com/Citytracking/toner/issues/34), [Issue 35](https://github.com/Citytracking/toner/issues/35). Version 3 will address remaining occational placement funk and overlap of marine labels.

* Added in **more city labels** in zooms 9, 10, and 11 from Dymo [Issue 15](https://github.com/Citytracking/toner/issues/15), [Issue 1](https://github.com/Citytracking/toner/issues/1), [Issue 51](https://github.com/Citytracking/toner/issues/51), [Issue 29](https://github.com/Citytracking/toner/issues/51), [Issue 27](https://github.com/Citytracking/toner/issues/27)

* Added **new park labels progressively** per the zoom. [Issue 13](https://github.com/Citytracking/toner/issues/13), [Issue 42](https://github.com/Citytracking/toner/issues/42).


mapnik/mapnik-render.py
import sys
import glob
import os.path
import cairo
import pyproj
import PIL.Image
import ModestMaps
import optparse

try:
    import mapnik
except ImportError:
    import mapnik2 as mapnik

optparser = optparse.OptionParser(usage="""%prog [options]
""")

defaults = {
    'fonts': 'fonts',
    'stylesheet': 'style.xml',
    'location': (37.804325, -122.271169),
    'zoom': 10,
    'size': (1024, 768),
    'output': 'out.png'
}

optparser.set_defaults(**defaults)

optparser.add_option('-f', '--fonts', dest='fonts',
                     type='string', help='Directory name for fonts. Default value is "%(fonts)s".' % defaults)

optparser.add_option('-s', '--stylesheet', dest='stylesheet',
                     type='string', help='File name of mapnik XML file. Default value is "%(stylesheet)s".' % defaults)

optparser.add_option('-l', '--location', dest='location',
                     nargs=2, type='float', help='Latitude and longitude of map center. Default value is %.6f, %.6f.' % defaults['location'])

optparser.add_option('-z', '--zoom', dest='zoom',
                     type='int', help='Zoom level of rendered map. Default value is %(zoom)d.' % defaults)

optparser.add_option('-d', '--dimensions', dest='size',
                     nargs=2, type='int', help='Width and height of rendered map. Default value is %d, %d.' % defaults['size'])

optparser.add_option('-o', '--output', dest='output',
                     type='string', help='File name of rendered map. Default value is "%(output)s".' % defaults)

if __name__ == '__main__':

    opts, args = optparser.parse_args()

    try:
        fonts = opts.fonts
        stylesheet = opts.stylesheet
        zoom = opts.zoom
        output = opts.output
    
        center = ModestMaps.Geo.Location(*opts.location)
        dimensions = ModestMaps.Core.Point(*opts.size)
        format = output.split(".").pop().lower()
        
        assert zoom >= 0 and zoom <= 19
        assert format in ('png', 'jpg', 'svg', 'pdf', 'ps')
    
        for ttf in glob.glob(os.path.join(fonts, '*.ttf')):
            mapnik.FontEngine.register_font(ttf)

    except Exception, e:
        print >> sys.stderr, e
        print >> sys.stderr, 'Usage: python mapnik-render.py <fonts dir> <stylesheet> <lat> <lon> <zoom> <width> <height> <output jpg/png/svg/pdf/ps>'
        sys.exit(1)

    osm = ModestMaps.OpenStreetMap.Provider()
    map = ModestMaps.mapByCenterZoom(osm, center, zoom, dimensions)
    
    srs = {'proj': 'merc', 'a': 6378137, 'b': 6378137, 'lat_0': 0, 'lon_0': 0, 'k': 1.0, 'units': 'm', 'nadgrids': '@null', 'no_defs': True}
    gym = pyproj.Proj(srs)

    northwest = map.pointLocation(ModestMaps.Core.Point(0, 0))
    southeast = map.pointLocation(dimensions)
    
    left, top = gym(northwest.lon, northwest.lat)
    right, bottom = gym(southeast.lon, southeast.lat)
    
    map = mapnik.Map(dimensions.x, dimensions.y)
    mapnik.load_map(map, stylesheet)
    map.zoom_to_box(mapnik.Envelope(left, top, right, bottom))
    
    img = mapnik.Image(dimensions.x, dimensions.y)
    
    # http://brehaut.net/blog/2010/svg_maps_with_cairo_and_mapnik
    # http://trac.mapnik.org/wiki/MapnikRenderers
    if format in ('svg', 'pdf', 'ps' ) :
        f = open(output, 'w')
        if format == 'svg' :
            surface = cairo.SVGSurface(f.name, dimensions.x, dimensions.y)
        elif format == 'pdf' :
            surface = cairo.PDFSurface(f.name, dimensions.x, dimensions.y)
        else :
            surface = cairo.PSSurface(f.name, dimensions.x, dimensions.y)
        context = cairo.Context(surface)
        mapnik.render(map, context)
        surface.finish()
    else :
        mapnik.render(map, img)
    
    	img = PIL.Image.fromstring('RGBA', (dimensions.x, dimensions.y), img.tostring())    	
    
        if format == 'jpg':
            img.save(output, quality=85)
        else :
            img.save(output)